v2/spanner-to-sourcedb/src/main/java/com/google/cloud/teleport/v2/templates/dbutils/dml/CassandraTypeHandler.java

```diff
@@ -0,0 +1,466 @@
+     * Generates a Type based on the provided {@link CassandraTypeHandler}.
+     *
+     * @param colName - which is used to fetch Key from valueJSON.
+     * @param valuesJson - contains all the key value for current incoming stream.
+     *
+     * @return a {@link Double} object containing the value represented in cassandra type.
+     */
+    public static Double handleCassandraDoubleType(String colName, JSONObject valuesJson) {
+        return valuesJson.getBigDecimal(colName).doubleValue();
+    }
+
+    /**
+     * Generates a Type based on the provided {@link CassandraTypeHandler}.
+     *
+     * @param colName - which is used to fetch Key from valueJSON.
+     * @param valuesJson - contains all the key value for current incoming stream.
+     *
+     * @return a {@link ByteBuffer} object containing the value represented in cassandra type.
+     */
+    public static ByteBuffer handleCassandraBlobType(String colName, JSONObject valuesJson) {
+        Object colValue = valuesJson.opt(colName);
+        if (colValue == null) {
+            return null;
+        }
+        return parseBlobType(colName, colValue);
+
+    }
+
+    /**
+     * Generates a Type based on the provided {@link CassandraTypeHandler}.
+     *
+     * @param colName - which is used to fetch Key from valueJSON.
+     * @param colValue - contains all the key value for current incoming stream.
+     *
+     * @return a {@link ByteBuffer} object containing the value represented in cassandra type.
+     */
+    private static ByteBuffer parseBlobType(String colName, Object colValue) {
+        byte[] byteArray;
+
+        if (colValue instanceof byte[]) {
+            byteArray = (byte[]) colValue;
+        } else if (colValue instanceof String) {
+            byteArray = java.util.Base64.getDecoder().decode((String) colValue);
+        } else {
+            throw new IllegalArgumentException("Unsupported type for column " + colName);
+        }
+
+        return ByteBuffer.wrap(byteArray);
+    }
+
+    /**
+     * Generates a Type based on the provided {@link CassandraTypeHandler}.
+     *
+     * @param colName - which is used to fetch Key from valueJSON.
+     * @param valuesJson - contains all the key value for current incoming stream.
+     *
+     * @return a {@link Date} object containing the value represented in cassandra type.
+     */
+    public static Date handleCassandraDateType(String colName, JSONObject valuesJson) {
+        return handleCassandraGenericDateType(colName, valuesJson, "yyyy-MM-dd");
+    }
+
+    /**
+     * Generates a Type based on the provided {@link CassandraTypeHandler}.
+     *
+     * @param colName - which is used to fetch Key from valueJSON.
+     * @param valuesJson - contains all the key value for current incoming stream.
+     *
+     * @return a {@link Date} object containing timestamp as value represented in cassandra type.
+     */
+    public static Date handleCassandraTimestampType(String colName, JSONObject valuesJson) {
```

Comment:
Similar to https://github.com/GoogleCloudPlatform/DataflowTemplates/pull/2054/files#r1877268541,
Would `Instant` work better?

Reply:
The `handleCassandraTimestampType` is now returning `LocalDate`.
As per the [driver docs](https://javadoc.io/static/com.datastax.oss/java-driver-core/4.13.0/com/datastax/oss/driver/api/core/data/SettableById.html) `setInstant` works by default for TimeStamp type.

Reply:
@VardhanThigle  we are doing this while assigning this value to cassandra DML Statement

Reply:
Would you be creating the DML as a query string or would you use the driver to set the values into a statement?
I would strongly lean on setting the values as it helps avoid un-intended sql injections and overall helps us make best use of Cassandra driver's type safety.

I don't have complete context on how we might do this when we actually write to Cassandra, to help ease our conversation, here's an example of a possible write - You would definitely have more elaborate ways that supply you the column names and values than the example here. 
## Example
### Table
Imagine we have a table like:
 ```sql
 CREATE TABLE IF NOT EXISTS example.data_types
                    id int PRIMARY KEY, 
                    time_col time, 
                    timestamp_col timestamp,
                    date_col date,
                    integer_col int,
                    medium_int_col int, 
                    tiny_int_col tinyint; 
```
### Java Example
```java 
        // Create an immutable list of column names
        List<String> columnNames = ImmutableList.of("id", "time_col", "timestamp_col", "date_col", "integer_col", "medium_int_col", "tiny_int_col");

        // Join the column names with commas
        String columns = String.join(", ", columnNames);           // Columns
        
      String questionMarks = columnNames.stream()
                .map(name -> "?")
                .collect(Collectors.joining(", "));
            // Prepare the statement
            PreparedStatement prepared = session.prepare("INSERT INTO example.data_types (" + columns ") " +
                    "VALUES (" + questionMarks + ")");

            // Bind values and execute
            // Here the values need to be strongly typed, for example for `Date` we need `LocalDate`.
            BoundStatement bound = prepared.bind(
                    1,
                    LocalTime.now(),
                    Instant.now(),
                    LocalDate.now(),
                    123456789,
                    (short) 25000,  // Medium int value (using Short)
                    (byte) 120      // Tiny int value (using Byte)
            );
```

## Something we might want to avoid
We might want to avoid creating a query that directly fills in the values into the string, bypassing the drivers settable interface (which helps with safety)

Reply:
@VardhanThigle we are doing the same way , I guess it will be cleared in DML PR

Reply:
Thank you for sharing this detailed explanation @VardhanThigle . Our implementation is aligned with your suggestions which will be shared for review in our upcoming PRs. Appreciate it.

Reply:
Thanks for the confirmation on prepared statements @taherkl .
In case we go with prepared statements, we will need to map incoming records from spanner to the types supported by the Cassandra 4.0 driver API.

On the read path I have experienced that the driver throws codec mismatch exceptions even for a mismatch like reading a `medium int` as an `int` instead of a `short`. Likewise on the Date, Time and related types too.

That was the motivation behind few of the review comments to map `TimeStamp` to java `Instant` and the like.

In case you are planning to reconvert it to a compatible type before calling the prepared statement in a future PR, it might be better to get to a compatible class in this class itself (in the scope of current PR to the extent possible)

If we do type marshallings at multiple places, possible customer issues around loss of precision (for time related types) or overflow/undeflow etc get very hard to debug. To the extent possible let's keep the conversion from spanner to Cassandra type at limited places.

Reply:
Yes, @VardhanThigle The return types will be bonded directly to the prepared statements and there won't be another layer of conversion. We will take care of that. We want to short PRs and hence you may have many questions about it. Please review the future PRs, which shall clear most of the confusions and address your suggestions.

Reply:
Thanks for the confirmation.
I can see that for timestamp we are still returning `LocalDate`
 [ref](https://github.com/GoogleCloudPlatform/DataflowTemplates/pull/2054/commits/957b8866afe58d464a0ace86dde6e1062337a4e7#diff-690e435057a9227e15eccda34696e95bebf8b51afe6a0ac421227a392276dce1R261)
 
 There's the Cassandra driver doc that I have linked to in most of these discussions around types. It's [here](https://javadoc.io/static/com.datastax.oss/java-driver-core/4.13.0/com/datastax/oss/driver/api/core/data/SettableById.html)
 
 As per the driver interface [setInstant](https://javadoc.io/static/com.datastax.oss/java-driver-core/4.13.0/com/datastax/oss/driver/api/core/data/SettableById.html#setInstant-com.datastax.oss.driver.api.core.CqlIdentifier-java.time.Instant-) method is supported for the CQL timestamp type
 
 where as,
 [setLocalDate](https://javadoc.io/static/com.datastax.oss/java-driver-core/4.13.0/com/datastax/oss/driver/api/core/data/SettableById.html#setLocalDate-com.datastax.oss.driver.api.core.CqlIdentifier-java.time.LocalDate-) method is supported for CQL date type.
 
 In case we use the `LocalDate` type and call `setLocalDate` for timestamp type on CQL, I am afraid, the driver might throw a codec mismatch exception. Same applies to mediumInt (works with Java `Short`) and tinyint (works with Java `byte`) type.
 
 
 The request here is to map the types to the ones expected by the Cassandra 4.0 driver.

Reply:
@VardhanThigle as we metioned earlier we are handling some unboxing stuff while assigning the object to DML
For example: we are parsing it to integer and while assiging based on sourceType we are casting to smallint , tinyint etc. we are doing the same in terms of timestamp as well

Reply:
What is the advantage of converting it to LocalDate and then Instant in this case , why not use Instant directly 

<img width="712" alt="image" src="https://github.com/user-attachments/assets/d2517cee-cccb-4c9c-be50-b24dee73719f" />

Reply:
@shreyakhajanchi, we aimed to generalize the approach by converting date and time or timestamp to a LocalDate, which can later be converted as needed for convenience but as of I now I have address as per suggestion

Reply:
Thanks.  This has been addressed now.
Let's directly map spanner objects to the types that are expected by Cassandra Driver. This will reduce any debugging for type handling or precisions at a single place.

I have had a best effort look at the changes in this PR but please check that all the types are compliant with the driver specification before merging.